### ISSUE TITLE
Android: Execute ecb without clicking in the notification bar

### ISSUE THREAD

**Issue by lereibol** [phonegap-build/PushPlugin#473]
Hi,

I have an app based on push notifications in Android and I have some doubts about the implementation. I would like to know if it is possible to execute the ecb function without clicking in the received notification. 

Thanks in advance,

**Comment by ajoyatgithub** [comment]
Are you considering a situation where the notification is discarded by the user, but you still need the parameters to reach the App?

**Comment by uriva** [comment]
+1
I have a similar sitiuation - on some messages sent by the server I want the app to perform something in the background without alerting the user at all.

**Comment by ajoyatgithub** [comment]
When a notification comes, it is sent to the App. This is the <a href="https://github.com/phonegap-build/PushPlugin/blob/master/src/android/com/plugin/gcm/GCMIntentService.java#L59">handler</a> that deals with recieved messages. A notification is sent if you 'createNotification'. If you need to send it to the app, 'PushPlugin.sendExtras'

**Comment by lereibol** [comment]
Hi everybody,

Thanks for your replies.

As Ajoy says, if you want to communicate with your app before showing the notification in the bar you must call to "sendExtras" method. 

If you realize, when the googleId is registered (No Notification is showed in the bar), the PushPlugin.java class calls to this method to send  the google registration number to your app, <a href="https://github.com/phonegap-build/PushPlugin/blob/master/src/android/com/plugin/gcm/PushPlugin.java#L76.">HERE</a>

So, as much as I know, it is required to change the  behavior of the plugin to get it, isn't it?

**Comment by ajoyatgithub** [comment]
Yes, the plugin behaviour must change. If the data from the server contains a key 'noAlert', a notification must not be created. But I am not sure how one could wake the app and send the data to it.

Or else it could be that the app receives all the cached extras whenever it is activated. I am not sure if the behaviour of <a href="https://github.com/phonegap-build/PushPlugin/blob/master/src/android/com/plugin/gcm/PushPlugin.java#L117">cached extras</a> is functional.

**Comment by lereibol** [comment]
Ok. Thanks :)

It's more or less what I was thinking.

What I do not know which is the best way to modify a plugin. I have already done some plugin samples but when I wanted modify one of them I had to remove from phonegap project and add it again. Is this the correct way to modify plugins?

Thanks in advance.

**Comment by ajoyatgithub** [comment]
Don't know if it's right. I do it the same way though. After modifying the plugin, I remove and re-add the plugin to my cordova project from my local fork.

**Comment by lon9man** [comment]
has anybody changed default behaviour of this plugin in question with 'noAlert' and etc?

**Comment by ajoyatgithub** [comment]
Not yet, probably. I have contributed this [answer][1] on StackOverflow. The person who asked the question said that the solution did work for him. I intend to submit a pull request soon, after testing it.

 [1]: http://stackoverflow.com/a/29437083/1761793

**Comment by lon9man** [comment]
ajoyatgithub, thanks for reply. my ideas about problem was the same.. but i think in a minute that phonegap in background can't process notifications.. but it works in background also.. 
need now:
1. iOS-version 
2. 'noAlert' (more simple then "1")
if you get updates please notify.. thanks..

**Comment by lon9man** [comment]
what about removing all native-notifications from notification bar from JS on some event (saying logout from my app)?

**Comment by ajoyatgithub** [comment]
I don't think this is also possible currently. Basically the application should create a notification and store it's id, and remove it when a condition is true

But there is this [line][1] that looks for a `notId` . I don't yet know what it its.

[1]: https://github.com/phonegap-build/PushPlugin/blob/master/src/android/com/plugin/gcm/GCMIntentService.java#L126

**Comment by lon9man** [comment]
it is possible with manual editing source and calling Android-code from JS using @Javascriptinterface. it works. now i remove all notifications on logout from my app.. in the same way we can remove only one needed notification (using notId).. but thats all need source editing. plugin functionality is not enough..
